### PR TITLE
Support for arbitrary data field names

### PR DESCRIPTION
**Pull request by bvanessen**
Added support in the generic data reader and synthetic data reader
clases for arbitrary data fields.  Added catch2 testing for the
synthetic data reader to check that arbitrary fields are found and
that there is checking for valid field names.

### CODE REVIEW

File: src/data_readers/data_reader_synthetic.cpp [hunk 1]
```diff
@@ -77,6 +77,28 @@ data_reader_synthetic::data_reader_synthetic(int num_samples,
...
+                                             std::map<data_field_type, std::vector<int>> data_fields,
+                                             bool shuffle)
+  : generic_data_reader(shuffle),
+    m_num_samples(num_samples),
+    m_synthetic_data_fields(data_fields)
+{
+  for(auto const& [data_field, dims] : data_fields) {
+    set_has_data_field(data_field, true);
+  }
+}
```
Comment (benson31) [inline src/data_readers/data_reader_synthetic.cpp:90]:
```suggestion
    m_synthetic_data_fields(std::move(data_fields))
{
  for (auto const& [data_field, dims] : m_synthetic_data_fields) {
    set_has_data_field(data_field, true);
  }
}
```

File: src/data_readers/unit_test/data_reader_synthetic_test_public_api.cpp [hunk 2]
```diff
@@ -187,3 +187,205 @@ TEST_CASE("Synthetic data reader public API tests",
...
+  //  std::set<std::string> active_data_fields = {"samples"};
+  auto s = GENERATE(range(1, 2));
+  El::Int num_samples = s;
+  std::set<lbann::data_field_type> data_fields = {"foo", "bar"};
+  std::map<lbann::data_field_type, std::vector<int>> fields;
+  int f = 0;
+  std::map<lbann::data_field_type, std::unique_ptr<lbann::CPUMat>>
+    owning_local_input_buffers;
+  std::map<lbann::data_field_type, lbann::CPUMat*> local_input_buffers;
+  for(auto data_field : data_fields) {
```
Comment (benson31) [inline src/data_readers/unit_test/data_reader_synthetic_test_public_api.cpp:216]:
```suggestion
  for (auto const& data_field : data_fields) {
```

File: src/data_readers/unit_test/data_reader_synthetic_test.cpp [hunk 3]
```diff
@@ -192,3 +200,58 @@ TEST_CASE("Synthetic data reader regression tests",
...
+    f++;
+  }
+
+  SECTION("fetch data field")
+  {
+    auto dr = std::make_unique<lbann::data_reader_synthetic>(num_samples,
+                                                             fields,
+                                                             false);
+    lbann::CPUMat X;
+    for (auto& [data_field, dims] : fields) {
```
Comment (benson31) [inline src/data_readers/unit_test/data_reader_synthetic_test.cpp:234]:
```suggestion
    for (auto const& [data_field, dims] : fields) {
```

File: src/data_readers/unit_test/data_reader_synthetic_test.cpp [hunk 4]
```diff
@@ -192,3 +200,58 @@ TEST_CASE("Synthetic data reader regression tests",
...
+  // Create a local copy of the RNG to check the synthetic data reader
+  lbann::fast_rng_gen ref_fast_generator;
+  ref_fast_generator.seed(lbann::hash_combine(42, 0));
+
+  auto s = GENERATE(range(1, 4));
+  El::Int num_samples = s;
+  std::vector<lbann::data_field_type> data_fields = {"foo", "bar"};
+  std::map<lbann::data_field_type, std::vector<int>> fields;
+  int f = 0;
+  for(auto data_field : data_fields) {
```
Comment (benson31) [inline src/data_readers/unit_test/data_reader_synthetic_test.cpp:222]:
```suggestion
  for (auto const& data_field : data_fields) {
```

File: src/data_readers/unit_test/data_reader_synthetic_test.cpp [hunk 5]
```diff
@@ -192,3 +200,58 @@ TEST_CASE("Synthetic data reader regression tests",
...
+    std::vector<int> dims = {s+f, s+f};
+    fields[data_field] = dims;
+    f++;
```
Comment (benson31) [inline src/data_readers/unit_test/data_reader_synthetic_test.cpp:225]:
```suggestion
    ++f;
```

File: src/data_readers/unit_test/data_reader_synthetic_test.cpp [hunk 6]
```diff
@@ -192,3 +200,58 @@ TEST_CASE("Synthetic data reader regression tests",
...
+    auto dr = std::make_unique<lbann::data_reader_synthetic>(num_samples,
+                                                             fields,
+                                                             false);
+    lbann::CPUMat X;
+    for (auto& [data_field, dims] : fields) {
+      std::cout << "Fetching " << data_field << std::endl;
+      X.Resize(dims[0] * dims[1], num_samples);
+
+      auto io_rng = lbann::set_io_generators_local_index(0);
+      for (auto j = 0; j < num_samples; j++) {
```
Comment (benson31) [inline src/data_readers/unit_test/data_reader_synthetic_test.cpp:239]:
```suggestion
      for (El::Int j = 0; j < num_samples; j++) {
```

File: src/data_readers/unit_test/data_reader_synthetic_test_public_api.cpp [hunk 7]
```diff
@@ -187,3 +187,205 @@ TEST_CASE("Synthetic data reader public API tests",
...
+    std::vector<int> dims = {s+f, s+f};
+    fields[data_field] = dims;
+    f++;
```
Comment (benson31) [inline src/data_readers/unit_test/data_reader_synthetic_test_public_api.cpp:219]:
```suggestion
    ++f;
```

File: src/data_readers/unit_test/data_reader_synthetic_test_public_api.cpp [hunk 8]
```diff
@@ -187,3 +187,205 @@ TEST_CASE("Synthetic data reader public API tests",
...
+    dr->set_mini_batch_size(num_samples);
+    dr->set_last_mini_batch_size(num_samples);
+    dr->set_initial_position();
+
+    dr->fetch(local_input_buffers, indices_fetched);
+
+    // Check all of the results that were fetched.  Ensure that the
+    // data fields are accessed in the same order that they are in the map
+    for (El::Int j = 0; j < num_samples; j++) {
+      for(auto data_field : data_fields) {
```
Comment (benson31) [inline src/data_readers/unit_test/data_reader_synthetic_test_public_api.cpp:250]:
```suggestion
      for (auto const& data_field : data_fields) {
```

File: src/data_readers/unit_test/data_reader_synthetic_test_public_api.cpp [hunk 9]
```diff
@@ -187,3 +187,205 @@ TEST_CASE("Synthetic data reader public API tests",
...
+    }
+  }
+
+  SECTION("fetch arbitrary bad data fields - no extra buffers")
+  {
+    std::map<lbann::data_field_type, std::vector<int>> test_fields;
+    std::map<lbann::data_field_type, lbann::CPUMat*> test_local_input_buffers;
+    lbann::data_field_type bad_field = "bar";
+    for(auto& data_field : data_fields) {
+      if(data_field != bad_field) {
```
Comment (benson31) [inline src/data_readers/unit_test/data_reader_synthetic_test_public_api.cpp:307]:
```suggestion
    for (auto const& data_field : data_fields) {
      if (data_field != bad_field) {
```

File: src/data_readers/unit_test/data_reader_synthetic_test_public_api.cpp [hunk 10]
```diff
@@ -187,3 +187,205 @@ TEST_CASE("Synthetic data reader public API tests",
...
+    dr->load();
+    dr->set_mini_batch_size(num_samples);
+    dr->set_last_mini_batch_size(num_samples);
+    dr->set_initial_position();
+
+    CHECK_THROWS(dr->fetch(local_input_buffers, indices_fetched));
+
+    // All data buffers should be empty since it will have thrown an exception
+    for (El::Int j = 0; j < num_samples; j++) {
+      for(auto data_field : data_fields) {
```
Comment (benson31) [inline src/data_readers/unit_test/data_reader_synthetic_test_public_api.cpp:291]:
```suggestion
      for (auto const& data_field : data_fields) {
```

File: src/data_readers/unit_test/data_reader_synthetic_test_public_api.cpp [hunk 11]
```diff
@@ -187,3 +187,205 @@ TEST_CASE("Synthetic data reader public API tests",
...
+      }
+    }
+  }
+
+  SECTION("fetch arbitrary bad data field with extra fields")
+  {
+    std::map<lbann::data_field_type, std::vector<int>> test_fields;
+    lbann::data_field_type bad_field = "bar";
+    for(auto& data_field : data_fields) {
+      if(data_field != bad_field) {
```
Comment (benson31) [inline src/data_readers/unit_test/data_reader_synthetic_test_public_api.cpp:271]:
```suggestion
    for (auto const& data_field : data_fields) {
      if (data_field != bad_field) {
```

File: src/data_readers/unit_test/data_reader_synthetic_test_public_api.cpp [hunk 12]
```diff
@@ -187,3 +187,205 @@ TEST_CASE("Synthetic data reader public API tests",
...
+
+    dr->fetch(test_local_input_buffers, indices_fetched);
+
+    // Check all of the results that were fetched.  Ensure that the
+    // data fields are accessed in the same order that they are in the map
+    for (El::Int j = 0; j < num_samples; j++) {
+      for(auto data_field : data_fields) {
+        std::cout << "checking " << data_field << std::endl;
+        auto& X = *(local_input_buffers[data_field]);
+        if(data_field == bad_field) {
```
Comment (benson31) [inline src/data_readers/unit_test/data_reader_synthetic_test_public_api.cpp:332]:
```suggestion
      for (auto const& data_field : data_fields) {
        std::cout << "checking " << data_field << std::endl;
        auto& X = *(local_input_buffers[data_field]);
        if (data_field == bad_field) {
```

File: src/data_readers/unit_test/data_reader_synthetic_test_public_api.cpp [hunk 13]
```diff
@@ -187,3 +187,205 @@ TEST_CASE("Synthetic data reader public API tests",
...
+    dr->setup(io_thread_pool->get_num_threads(), io_thread_pool.get());
+    dr->set_rank(0);
+    dr->set_comm(&comm);
+    dr->set_num_parallel_readers(1);
+    dr->load();
+    dr->set_mini_batch_size(num_samples);
+    dr->set_last_mini_batch_size(num_samples);
+    dr->set_initial_position();
+
+    for(auto& data_field : data_fields) {
```
Comment (benson31) [inline src/data_readers/unit_test/data_reader_synthetic_test_public_api.cpp:373]:
```suggestion
    for(auto const& data_field : data_fields) {
```

File: src/data_readers/unit_test/data_reader_synthetic_test_public_api.cpp [hunk 14]
```diff
@@ -187,3 +187,205 @@ TEST_CASE("Synthetic data reader public API tests",
...
+      dr->set_has_data_field(data_field, false);
+    }
+
+    CHECK_THROWS(dr->fetch(local_input_buffers, indices_fetched));
+
+    // All data buffers should be empty since it will have thrown an exception
+    for (El::Int j = 0; j < num_samples; j++) {
+      for(auto data_field : data_fields) {
```
Comment (benson31) [inline src/data_readers/unit_test/data_reader_synthetic_test_public_api.cpp:381]:
```suggestion
      for (auto const& data_field : data_fields) {
```

**Review by benson31 — APPROVED** [review]
Mostly style issues.

**Comment by bvanessen** [comment]
@benson31 Can you take a look at this again.  I have updated the testing.  I think that maybe we should merge this without adding support for the python data reader in this PR.

**Review by benson31 — APPROVED** [review]

**Review by benson31 — APPROVED** [review]
LGTM

**Pull request merged**